src/udp/socket.rs

```diff
@@ -518,264 +535,234 @@ quick_error! {
-        Bytes::from(encrypted)
+    pub fn new_ttl_incrementer(
+        handle: &Handle,
+        socket: WithAddress,
+        their_pk: PublicKey,
+        our_sk: SecretKey,
+        duration_to_reach_max_ttl: Duration,
+    ) -> HolePunching {
+        HolePunching {
+            socket: Some(socket),
+            sending_msg: None,
+            timeout: Timeout::new(Duration::new(0, 0), handle),
+            their_pk: their_pk,
+            our_sk: our_sk,
+            phase: HolePunchingPhase::Syn {
+                time_of_last_ttl_increment: Instant::now(),
+                ttl_increment_duration: duration_to_reach_max_ttl / MAX_TTL,
+            },
+        }
     }
 
-    /// Try to decrypt raw bytes into hole punching message.
-    fn decrypt_msg(&self, msg: &Bytes) -> Result<HolePunchMsg, SecureSerialiseError> {
-        secure_deserialise(msg, &self.their_pk, &self.our_sk)
-    }
+    fn flush(&mut self) -> Result<Async<()>, HolePunchError> {
+        loop {
+            match unwrap!(self.socket.as_mut()).poll_complete() {
+                Err(e) => return Err(HolePunchError::SendMessage(e)),
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Ok(Async::Ready(())) => (),
+            };
 
-    /// Sends hole punching message message to the given socket.
-    fn send_msg(
-        &self,
-        msg: &HolePunchMsg,
-        socket: WithAddress,
-    ) -> BoxFuture<WithAddress, HolePunchError> {
-        socket
-            .send(self.encrypt_msg(msg))
-            .map_err(HolePunchError::SendMessage)
-            .into_boxed()
-    }
+            if let Some(bytes) = self.sending_msg.take() {
+                match unwrap!(self.socket.as_mut()).start_send(bytes) {
+                    Err(e) => return Err(HolePunchError::SendMessage(e)),
+                    Ok(AsyncSink::Ready) => continue,
+                    Ok(AsyncSink::NotReady(bytes)) => {
+                        self.sending_msg = Some(bytes);
+                        return Ok(Async::NotReady);
+                    },
+                }
+            }
 
-    fn inc_ttl_and_syn_acks(&mut self, socket: WithAddress) -> io::Result<WithAddress> {
-        self.syns_acks_sent += 1;
-        if self.syns_acks_sent % 5 == 0 && self.ttl_increment != 0 {
-            let ttl = socket.ttl()?;
-            socket.set_ttl(ttl + self.ttl_increment)?;
+            return Ok(Async::Ready(()));
         }
-        Ok(socket)
     }
 
-    /// Creates timeout of 200 miliseconds from when this function is called.
-    /// Reduces boilerplate and allows to easily add some dynamic to timeouts.
-    fn msg_timeout(&self) -> Instant {
-        Instant::now() + Duration::from_millis(200)
+    fn send_msg(&mut self, msg: &HolePunchMsg) {
+        let encrypted = unwrap!(secure_serialise(msg, &self.their_pk, &self.our_sk));
+        let bytes = Bytes::from(encrypted);
+        assert!(self.sending_msg.is_none());
```

Comment:
Hmm, do we really want to panic here?

Reply:
Yes, you shouldn't be able to get to this part of the code with a `None` unless there's a bug. I could make it a `debug_assert` instead though...